pkg/ring/util.go

```diff
@@ -82,6 +82,38 @@ func WaitInstanceState(ctx context.Context, r *Ring, instanceID string, state In
 	return backoff.Err()
 }
 
+// WaitRingStability monitors the ring topology for the provided operation and waits until it
+// keeps stable for at least minStability.
+func WaitRingStability(ctx context.Context, r *Ring, op Operation, minStability, maxWaiting time.Duration) error {
+	const pollingFrequency = time.Second
+
+	// Configure the max waiting time as a context deadline.
```

Comment:
I'd suggest removing `maxWaiting` and rely on context only. Client calling this can already setup context deadline, if needed. The code right now returns `context.DeadlineExceeded` when `maxWaiting` is reached, which client may confuse with its own context deadline.

Reply:
I thought the same, but using it from the caller perspective is more annoying, because it's not something you can do inline. You need to get a new context, defer the cancel(), call the WaitRingStability().